src/server/page_server.cpp

```diff
@@ -944,6 +913,130 @@ page_server::disconnect_all_tran_servers ()
+}
+
+void
+page_server::execute_catchup (cubthread::entry &entry, const LOG_LSA catchup_lsa)
+{
+  // Initialize variables for buffers
+  constexpr int PAGE_CNT_IN_BUFFER = 10;       // Arbitrarily chosen
+  const size_t buffer_size_in_total = PAGE_CNT_IN_BUFFER * LOG_PAGESIZE + MAX_ALIGNMENT;
+  auto log_pgbuf_buffer1 = std::unique_ptr<char []> { new char[buffer_size_in_total] };
+  auto log_pgbuf_buffer2 = std::unique_ptr<char []> { new char[buffer_size_in_total] };
+  char *const aligned_log_pgbuf_buffer1 = PTR_ALIGN (log_pgbuf_buffer1.get (), MAX_ALIGNMENT);
+  char *const aligned_log_pgbuf_buffer2 = PTR_ALIGN (log_pgbuf_buffer2.get (), MAX_ALIGNMENT);
+
+  auto log_pgptr_vec = std::vector<LOG_PAGE *> {};
+  auto log_pgptr_recv_vec = std::vector<LOG_PAGE *> {};
+
+  for (int i = 0; i < PAGE_CNT_IN_BUFFER; i++)
+    {
+      log_pgptr_vec.emplace_back (reinterpret_cast<LOG_PAGE *> (aligned_log_pgbuf_buffer1 + i * LOG_PAGESIZE));
+      log_pgptr_recv_vec.emplace_back (reinterpret_cast<LOG_PAGE *> (aligned_log_pgbuf_buffer2 + i * LOG_PAGESIZE));
+    }
+
+  const LOG_PAGEID start_pageid = log_Gl.hdr.append_lsa.pageid;
+  const LOG_PAGEID end_pageid = catchup_lsa.offset == 0 ? catchup_lsa.pageid - 1 : catchup_lsa.pageid;
+  const size_t total_page_count = end_pageid - start_pageid + 1;
+
+  // Request pages to the followee
+  auto request_pages_to_buffer = [this, &log_pgptr_recv_vec] (LOG_PAGEID start_pageid, size_t request_page_cnt) -> int
+  {
+    auto lockg = std::lock_guard <std::mutex> (m_followee_conn_mutex);
+    if (m_followee_conn == nullptr)
+      {
+	// It has been disconnected somewhere, because of an error or server shutdown.
+	er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED, 0);
+	return ER_CONN_PAGE_SERVER_CANNOT_BE_REACHED;
+      }
+    const int error_code = m_followee_conn->request_log_pages (start_pageid, request_page_cnt, log_pgptr_recv_vec);
+    if (error_code != NO_ERROR)
+      {
+	/* There are two kinds of errors either from client-side or server-side.
+	 * client-side error: it fails to send or receive.
+	 * server-side error: the followee fails to fetch log pages. For example, due to
+	 *  - the followee PS any truncate some log pages requested.
+	 *  - some requested pages are corrupted
+	 *  - or something
+	 * TODO We have to handle them. Probably, we should change the followee to catch up with through ATS.
+	 */
+	assert_release (false);
+      }
+    return error_code;
+  };
+
+  _er_log_debug (ARG_FILE_LINE, "[CATCH-UP] The catch-up starts with pages ranging from %lld to %lld, count = %lld.\n",
+		 start_pageid, end_pageid, total_page_count);
+
+  LOG_PAGEID request_start_pageid = start_pageid;
+  size_t remaining_page_cnt = total_page_count;
+  size_t request_page_cnt = std::min (log_pgptr_recv_vec.size (), remaining_page_cnt);
+  auto req_future = std::async (std::launch::async, request_pages_to_buffer, request_start_pageid, request_page_cnt);
```

Comment:
I understand that `std::async` is to parallelize:
- requests to followee (producing log pages)
- and appending in follower (consume those)
Having `execute_catchup` run as an async task, while also making use of `std::async` which - behind the scene, creates another thread pool - somewhat defeats the purpose.

What I would do is:
- implement these two aspects as two proper separate implementations of `cubthread::task` - as the async responder does
- move the code in a separate file
- have these two tasks pass the received pages between them as done here
- when  the "produce" task is done, it signals this to the "consume" task
- when the "consume" task is done, it does the signalling

Another, less important aspect, the "swap" approach assumes that the "consume" is much faster than "produce" which is probably true 99.(9)% of the time.

Reply:
> Having execute_catchup run as an async task, while also making use of std::async which - behind the scene, creates another thread pool - somewhat defeats the purpose.

I don't understand this. Do you mean this parallel approach is not worthy? Or, does it defeat the purpose to use `std::async`? I guess it's worth having two jobs parallel, though: appending 1\~10 pages into its local buffer, while pulling next pages, 11\~20, from remote. Are you against it? or just against using `std::async` instead of CUBRID's `task?

Reply:
The approach with 2 separate threads is definitely ok.
But the implementation is cumbersome with `std::async` because it mixes too many concepts together and it just does not look "clean" to my eyes.

It will certainly work.

However there can be a small performance bottleneck as the log page retrieval in the `std::async` is not completely independent of the actual `page_server::execute_catchup` execution - both threads have a "piece-meal" way of progressing. So, supposing that retrieval goes way faster than processing (very unlikely) the retrieval will be starved.

Take this comment as a "nit" - we can certainly change later.

Reply:
I've got your point. I've chosen the most handy one at that time, but yes I relate to that it gets complicated with many concepts. I'll refactor it. Thank you.
I'll gather probable refactoring in http://jira.cubrid.org/browse/LETS-796, which I'll return after making progress.